router/batchrouter/asyncdestinationmanager/snowpipestreaming/snowpipestreaming_test.go

```diff
@@ -310,6 +344,97 @@ func TestSnowpipeStreaming(t *testing.T) {
 			"status":        "failed",
 		}).LastValue())
 	})
+
+	t.Run("Upload with unauthorized schema error should add backoff", func(t *testing.T) {
+		statsStore, err := memstats.New()
+		require.NoError(t, err)
+
+		sm := New(config.New(), logger.NOP, statsStore, destination)
+		sm.channelCache.Store("RUDDER_DISCARDS", rudderDiscardsChannelResponse)
+		sm.api = &mockAPI{
+			createChannelOutputMap: map[string]func() (*model.ChannelResponse, error){
+				"USERS": func() (*model.ChannelResponse, error) {
+					return &model.ChannelResponse{Code: internalapi.ErrSchemaDoesNotExistOrNotAuthorized}, nil
+				},
+			},
+		}
+		managerCreatorCallCount := 0
+		sm.managerCreator = func(_ context.Context, _ whutils.ModelWarehouse, _ *config.Config, _ logger.Logger, _ stats.Stats) (manager.Manager, error) {
+			sm := snowflake.New(config.New(), logger.NOP, stats.NOP)
+			managerCreatorCallCount++
+			mockManager := newMockManager(sm)
+			mockManager.createSchemaErr = fmt.Errorf("failed to create schema")
+			return mockManager, nil
+		}
+		mockClock := &mockClock{}
+		mockClock.nowTime = sm.now()
+		sm.authzBackoff = newAuthzBackoff(time.Second*10, mockClock)
+		asyncDestStruct := &common.AsyncDestinationStruct{
+			Destination: destination,
+			FileName:    "testdata/successful_user_records.txt",
+		}
+		require.Equal(t, false, sm.authzBackoff.isInBackoff())
+		output1 := sm.Upload(asyncDestStruct)
+		require.Equal(t, 2, output1.FailedCount)
+		require.Equal(t, 0, output1.AbortCount)
+		require.Equal(t, 1, managerCreatorCallCount)
+		require.Equal(t, true, sm.authzBackoff.isInBackoff())
+
+		sm.Upload(asyncDestStruct)
+		// client is not created again due to backoff error
+		require.Equal(t, 1, managerCreatorCallCount)
+		require.Equal(t, true, sm.authzBackoff.isInBackoff())
+
+		mockClock.nowTime = sm.now().Add(time.Second * 5)
+		require.Equal(t, true, sm.authzBackoff.isInBackoff())
+		mockClock.nowTime = sm.now().Add(time.Second * 20)
+		require.Equal(t, false, sm.authzBackoff.isInBackoff())
+
+		sm.Upload(asyncDestStruct)
+		// client created again since backoff duration has been exceeded
+		require.Equal(t, 2, managerCreatorCallCount)
+		require.Equal(t, false, sm.authzBackoff.isInBackoff())
```

Comment:
Since here managerCreator will return an error right, sm.authzBackoff.isInBackoff() should return True right, because Post upload backoff will be set.

Reply:
Updated. The value was coming as `true` because of the difference in the clock implementation in the two backoff approaches